Fix #30, bug where the package cannot properly parse a file that contains Markdown code blocks with front matter in them

caendesilva opened the pull request:
This PR fixes #30 

The PR is fully backwards compatible as it does not change the existing parser, but instead adds a new one: `YamlFrontMatter::markdownCompatibleParse()` which uses code by https://github.com/eklausme (I hope this is okay, I'm happy to add you as a co-author if you see this!

Before:
![image](https://user-images.githubusercontent.com/95144705/160793264-15d59ea1-44a1-4019-9b01-90e41b06aa75.png)

After:
![image](https://user-images.githubusercontent.com/95144705/160793131-0905fb1a-b67b-4e25-b790-b6f7086c70fb.png)

The source markdown used for the example: (with the inline backticks replaced)
```markdown
---
title: test
foo: bar
---

A paragraph in a Markdown Post.
This file contains a Markdown code block that the original parser does not handle.
See https://github.com/spatie/yaml-front-matter/discussions/30.

'''
---
title: The Title of Markdown Code
---

A paragraph in Markdown Code
'''

```

freekmurze commented on src/YamlFrontMatter.php:
```diff
@@ -23,6 +23,45 @@ public static function parse(string $content): Document
+     * @param string $content
+     * @return Document
+     */
+    public static function markdownCompatibleParse(string $content): Document
```
This code does not meet our quality expectations.

- Do not use abbreviations
- Always use strict checking
- `if` blocks should always have a body
- if you have to use a comment, there's a high change it could be replace with a function with a clear name

Could you refactor this code to be as readable as possible?

caendesilva replied:
Yeah I will give it a shot! Can't promise I'll be able to get it done today though.

caendesilva commented:
Okay I refactored the code to use an improved version of what I used for the Hyde framework. I'm converting everything to arrays, which may be a bit inefficient but I think it makes it much more readable and understandable. I also added two more tests. Let me know if its okay or if I should change anything else. 

This is my first community PR so I'd love to get feedback for the many more I hope to make :)

freekmurze commented on src/YamlFrontMatter.php:
```diff
@@ -23,6 +23,79 @@ public static function parse(string $content): Document
+                if (sizeof($frontMatterControlBlockIndex) === 0) {
+                    // This is the first control block
+                    $frontMatterControlBlockIndex['start'] = $lineNumber;
+                } elseif (sizeof($frontMatterControlBlockIndex) === 1) {
```
Could you refactor this so we don't use `elsif` and `break`?

caendesilva replied:
Done.

freekmurze commented on src/YamlFrontMatter.php:
```diff
@@ -23,6 +23,79 @@ public static function parse(string $content): Document
+     * @param string $content
+     * @return Document
+     */
+    public static function markdownCompatibleParse(string $content): Document
```
This whole function is kinda long, could you split it up in smaller logical parts?

caendesilva replied:
I'll refactor it! I think I will separate it all into its own class as I think it will be easier that way.

*This comment is in reply to all three notations

caendesilva replied:
Created a new class, [`ComplexMarkdownParser`](https://github.com/spatie/yaml-front-matter/blob/2176038f1dadefc219d4fc29eca034024f222b3f/src/ComplexMarkdownParser.php) as I felt it would be easier (and make more sense) to work in an object oriented matter. I kept the static `markdownCompatibleParse` function but changed it to offload the logic into the new class.

```php
public static function markdownCompatibleParse(string $content): Document
{
    return (new ComplexMarkdownParser($content))->parse();
}
```

The class still has 100% test coverage (according to PHPStorm) and I also continuously ran smoke tests to ensure that the actual output did not change.

caendesilva commented on src/YamlFrontMatter.php:
```diff
@@ -23,6 +23,79 @@ public static function parse(string $content): Document
+        $frontMatterControlBlockIndex = [];
+        foreach ($lines as $lineNumber => $lineContents) {
+            // If the line starts with three dashes, it's a front matter control block
+            if (substr($lineContents, 0, 3) === '---') {
```
Thank you for the tip! Added it in https://github.com/spatie/yaml-front-matter/blob/3616b8034dc160036245636a61f2d5976d3a8e4a/src/ComplexMarkdownParser.php#L107

caendesilva commented:
Let me know if there is anything else I can improve!

Also, @freekmurze, I really want to thank you for taking the time to give me such great feedback and helping me become a better developer! This is my first time contributing to a community package and thanks to you I feel like I have learned a lot!

freekmurze commented on src/ComplexMarkdownParser.php:
```diff
@@ -0,0 +1,168 @@
+
+use Symfony\Component\Yaml\Yaml;
+
+/**
```
The class will become some much lighter if we remove all the comments. Using good names will already make clear what is going on.

Could you remove all the comments?

caendesilva replied:
Will do! Fill fix the other conversations as well!

freekmurze commented on src/ComplexMarkdownParser.php:
```diff
@@ -0,0 +1,168 @@
+    public function __construct(string $content)
+    {
+        $this->content = $content;
+        $this->lines = explode("\n", $this->content);
```
Use single quotes by default.

caendesilva replied:
Thank you for the reminder. Here (as far as my understanding of PHP strings goes) we need the double quotes so that the newline character `\n` gets evaluated. With the double quotes, the parser works for inputs using both `LF` and `CRLF `documents. Is there a better way to do this?

I wrote two new tests, to test LF/CRLF line endings and tried switching to using `PHP_EOL`, but the tests failed as my Windows environment could then not parse the LF sequence. I could write a helper method to detect the actual file endings used in each document, but I'm not sure if that solves the issue or just adds more complexity.

freekmurze commented on src/ComplexMarkdownParser.php:
```diff
@@ -0,0 +1,168 @@
+     * @param string $line
+     * @return bool
+     */
+    private function isFrontMatterControlBlock(string $line): bool
```
We use `protected` by default instead of `private`. Change all uses of `private` to `protected`.

Additionally, this helper function is only being used only later in the code. Move this function underneath the code where it is being used.

freekmurze commented on src/ComplexMarkdownParser.php:
```diff
@@ -0,0 +1,168 @@
+        if (!isset($this->frontMatterStartLine)) {
+            $this->frontMatterStartLine = $lineNumber;
+            return;
+        }
```
Add an empty line between the two if block to make the code breathe.

freekmurze replied:
Add an empty line between the two if block to make the code breathe.

freekmurze commented on src/ComplexMarkdownParser.php:
```diff
@@ -0,0 +1,168 @@
+                }
+            }
+        }
+        return implode("\n", $matter);
```
Use single quotes.

freekmurze replied:
Use single quotes.

caendesilva commented on src/ComplexMarkdownParser.php:
```diff
@@ -0,0 +1,168 @@
+     * @param string $line
+     * @return bool
+     */
+    private function isFrontMatterControlBlock(string $line): bool
```
Replacing scope and moving all helpers to the bottom, in the order they are first used.

caendesilva commented:
Okay, pushing the commits now. I made it all in separate commits as I think that makes it easier to follow. I assume you will probably squash them before merging. As you can see I did not change the double quotes to single quotes as I wanted to await your input on the concern I placed as a comment in the first change conversation.

caendesilva commented:
That's super cool! Thank you!
